### ISSUE TITLE
Retroactively including a hashtag in a post doesn't seem to propagate the post to instances that follow the hashtag but not the user

### ISSUE THREAD

**Issue by JackTheFoxOtter** [mastodon/mastodon#25537]
### Steps to reproduce the problem

1. Have two users A and B on two different instances
2. User A doesn't follow user B, but they follow the hashtag (at)Example
3. User B makes a post that includes the hashtag initially, and it shows up in user A's feed
4. User B makes a post that doesn't include the hashtag initially and then edits the post to include it. That post doesn't show up in user A's feed.

### Expected behaviour

Posts including a hashtag I follow should show up in my feed provided my home instance knows the instance it comes from

### Actual behaviour

Posts including the hashtag only show up if it was present in the initial version of the post, not if it has been added retroactively through editing

### Detailed description

I noticed that for some users, some posts they made with a hashtag I follow didn't show up in my home instance's feed, but others did. Some digging revealed that all posts that didn't show up seem to have been edited and had the hashtag retroactively added.

I suspect this only happens when noone on the instance of user A follows user B, as otherwise the instance would be aware of all posts they make. I don't have a test account to confirm that though.

### Specifications

Mastodon v4.1.2 on Chrome 114.0.5735.134

**Comment by Cassolotl** [comment]
Okay, so, my understanding of how federation works is based on [this diagram](https://commons.wikimedia.org/wiki/File:Mastodon_timelines.png).

I *think* that if no one on Instance A is following User B, nothing they post with a hashtag will show up in hashtag searches on Instance A, ***unless*** someone *followed by* User A on Instance A boosts User B's post. Then that one post reaches Instance A, and it'll show up in the federated timeline and hashtag searches. (Searching a specific post based on the URL also brings a post into the instance's known fediverse and thus the federated timeline and hashtag searches.)

To put it another way, the hashtag search is *just* a hashtag search of the federated timeline. The hashtag search can't pull in posts that aren't posted or boosted by accounts not followed by people on the instance. If someone boosts a post then a post might make it into the federated timeline (and therefore into hashtag searches), but it'd just be that one post. (Searching a specific post based on the URL also brings a post into the instance's known fediverse and thus the federated timeline and hashtag searches.)

So in step 3 in your post above with the list of reproduction steps, the only way it would have worked out that way is:
1. if someone on Instance A unfollowed User B between steps 3 and 4, or
2. if someone searched for the first post using a URL, or
3. if someone followed by someone on Instance A boosted just the first post.

That's my understanding of it anyway. Maybe someone else can step in and verify?

**Comment by JackTheFoxOtter** [comment]
Based on the image, that might be correct, but that doesn't really match my observations.
The posts weren't boosted, and since it was over a span of multiple days with some posts showing up and others not,
I doubt they were continuously followed or unfollowed.
I did however find one post that didn't show up in my timeline which wasn't edited and had the hashtag, so I think my own theory might also be flawed (although most of the ones that didn't show up were edited, and the ones that did show up weren't, although I suppose perhaps this could have been just been by chance and unrelated).

I think the manually searching for the link one is the most likely one, as I tested that, and that seems to just make the instance be aware of the post and it shows up in my feed upon reloading the page. But when a link is shared online, I kind of feel like it's more likely someone actually follows the link than copying it into the search bar of their mastodon instance.

It does feel extremely unreliable and hard to understand as an end user though. What's the point of using hashtags if they don't actually influence the discoverability? I though the ideas was using those to find likeminded people you don't already know / follow, so it perhaps is an oversight that hashtags don't propagate accross instances by themselves.

**Comment by Cassolotl** [comment]
>It does feel extremely unreliable and hard to understand as an end user though. What's the point of using hashtags if they don't actually influence the discoverability? I though the idea was using those to find likeminded people you don't already know / follow, so it perhaps is an oversight that hashtags don't propagate accross instances by themselves.

Yes, it does. Federated social media is sort of a bit like that by nature, which is a horrible thing to have to tell people who are just trying to have a nice time easily, and I can only apologise for that. 😬 

>Also, I think the terminology here adds massively to the confusion.
>"Follow User" means all posts they do show up in your feed.
>"Follow Hashtag" apparently means nothing then, it's just a filter on your home instance, but it isn't called that...

That's not quite right - if you follow the hashtag then you see (in your home timeline) any post that your instance knows about that contains that hashtag, regardless of whether or not you follow them. If you're on a single-user instance then it would be a Home timeline filter, but it's actually a federated timeline filter.

The federated timeline contains:
* Any post by anyone you follow
* Any post boosted by anyone you follow
* Any post by anyone that anyone on your instance follows
* Any post boosted by anyone that anyone on your instance follows
* Any post searched in the web UI search using its URL
* There are more ways to get into the federated timeline that I can't remember right now!

So, if you're on an instance with a lot of people, that's a LOT of posts, many more than would just appear in your home timeline. And any of those that contain a hashtag will show up in your Home timeline if you follow that hashtag, even if you personally don't follow the user who posted it.

>I think the manually searching for the link one is the most likely one, as I tested that, and that seems to just make the instance be aware of the post and it shows up in my feed upon reloading the page.

Yes, if you had searched for a post using the URL (or done something else that would make a post known to your instance), that would definitely scupper the replication test you listed in your first post here. In order to truly test this bug, you would have to have User A and User B not follow each other, **and** you would need to **guarantee** that no one on User A's instance was following someone who had boosted User B.

**Comment by JackTheFoxOtter** [comment]
I wonder if there would be any significant challenge in propagating posts with followed hastags similarly between instances as posts / boosts by followed people are propagated. Basically treating following hastags similarly to following people.

In any case thank you very much for your explanation. I don't want to create test users on public instances, but I might spin up my own Docker containers to test this behavior further if I find the time. I'm still new to properly understanding how federation / Mastodon works in detail, but I do think there would be a lot of value in subscribing to hashtags accross instances.

I'm on a reasonably sized instance myself, but the community I am part of is pretty small, and scattered accross many different instances. I didn't realize up until yesterday that following my community hashtag doesn't guarantee actually seeing all posts of the community, which I find pretty unfortunate. I'm thinking about creating a little script that just uses the "search by URL" feature to synchronize posts with the hashtag across a couple of instances now, but that's pretty hacky and I think a native solution would be a lot more elegant.

**Comment by Cassolotl** [comment]
Yeah, basically hashtag-related communities rely on people boosting all the good posts to make sure they make it to as many instances as possible! The fediverse works best when people boost very liberally. :)

Anyway, you're welcome, and thanks for being interested!

**Comment by sfescape** [comment]
If you look at gup.pe (a specialized instance designed to implement 'group' posting), its behavior is very close to how I think people intrinsically think about hashtags.

Anyone, on any federated server, that @mentions a gup.pe group member gets their post boosted automatically to every server on which someone is following that gup.pe group member.

**Comment by JackTheFoxOtter** [comment]
That's good to know, however that very much feels like a workaround around the issue instead of a proper fix. I'm wondering if there would be significant downsides to extending Mastodon to also subscribe to followed hashtags between instances. I mean, it would probably lead to more traffic as for popular hashtags more posts would be distributed, but I wanna say that would probably even be a good thing. If anyone has some insights on this please do share, if my idea doesn't fit into Mastodon I'd like to understand why.

**Comment by brezelman** [comment]
Since you traced the problem back to the lack of federation of hashtags, this is probably a duplicate of #22637.
